src/IW/Sync/Update.hs

```diff
@@ -28,13 +31,68 @@ fetchAndUpsertRepos
        , WithError m
        )
     => m ()
-fetchAndUpsertRepos = do
-    gitHubRepos <- fetchAllHaskellRepos
+syncRepos = do
+    today <- liftIO getToday
+    sync syncReposByDate firstHaskellRepoCreated today 15 1
+
+syncReposByDate
+    :: forall env m.
+       ( MonadCabal m
+       , MonadUnliftIO m
+       , WithDb env m
+       , WithLog env m
+       , WithError m
+       )
+    => Day
+    -> Day
+    -> Int
+    -> m Int
+syncReposByDate from to page = do
+    gitHubRepos <- liftGithubSearchToApp $ searchHaskellReposByDate from to page
     let repos = map fromGitHubRepo gitHubRepos
     upsertRepos repos
-    mapConcurrently_ fetchAndUpdateCategories repos
+    mapConcurrently_ syncCategories repos
+    pure $ length repos
+
+syncCategories
+    :: forall env m.
+       ( MonadCabal m
+       , MonadUnliftIO m
+       , WithDb env m
+       , WithLog env m
+       , WithError m
+       )
+    => Repo
+    -> m ()
+syncCategories Repo{..} = do
+    categories <- getCabalCategories repoOwner repoName
+    updateRepoCategories repoOwner repoName categories
+
+sync
+    :: forall env m.
+       ( MonadCabal m
+       , MonadUnliftIO m
+       , WithDb env m
+       , WithLog env m
+       , WithError m
+       )
+    => (Day -> Day -> Int -> m Int) -- ^ The function used for synchronization
+    -> Day     -- ^ Oldest day that will be reached
+    -> Day     -- ^ Most recent day to start synchronization from
+    -> Integer -- ^ Interval of days for the date range to be split into
+    -> Int     -- ^ Page
+    -> m ()
+sync syncFunction oldest recent interval page =
+    if recent == oldest
+    then log I $ "Oldest day " <> julianDayToIso oldest <> " reached. Synchronization complete."
+    else do
+        resCount <- syncFunction intervalStart recent page
+        if resCount < 100
+            then sync syncFunction oldest nextRecent interval 1
+            else sync syncFunction oldest recent interval $ next page
   where
-    fetchAndUpdateCategories :: Repo -> m ()
-    fetchAndUpdateCategories Repo{..} = do
-        categories <- getCabalCategories repoOwner repoName
-        updateRepoCategories repoOwner repoName categories
+    intervalStart :: Day
+    intervalStart = negate interval `addDays` recent
+
+    nextRecent :: Day
+    nextRecent = pred intervalStart
```

Comment:
Regarding HLint warning on the CI: you can disable this warning in this particular file for now. You can see how it's done here:

* https://github.com/kowainik/kowainik.github.io/blob/0f889b8c85cccc5296bd1f4a5c7499f219d77c20/site.hs#L28

And, if you have time, you can also open the corresponding issue to the `relude` and we will fix it in the next release :slightly_smiling_face: